### PR TITLE
[dmd-cxx] Fix warnings when building with all enabled in GCC

### PR DESCRIPTION
**Pull request by ibuclaw**
In GDC tree, the default compiler warnings are:
```
-W -Wall \
-Wno-narrowing \
-Wwrite-strings \
-Wcast-qual \
-Wno-long-long \
-Wno-variadic-macros \
-Wno-overlength-strings \
-fno-common
```
And no complaints come from the compiler.

In DMD tree, building with `ENABLE_WARNINGS=1` is now mostly silent, save for the backend and other non-shared frontend sources.

### CODE REVIEW

File: src/ctfeexpr.c [hunk 1]
```diff
@@ -1562,7 +1560,7 @@ Expression *ctfeIndex(Loc loc, Type *type, Expression *e1, uinteger_t indx)
         StringExp *es1 = (StringExp *)e1;
         if (indx >= es1->len)
         {
-            error(loc, "string index %llu is out of bounds [0 .. %llu]", indx, (ulonglong)es1->len);
+            error(loc, "string index %llu is out of bounds [0 .. %llu]", (ulonglong)indx, (ulonglong)es1->len);
             return CTFEExp::cantexp;
```
Comment (dkgroot) [inline src/ctfeexpr.c]: @ibuclaw 
For size_t related elements it is better to use %zu / %zd etc.
ie:
```
size_t x = ...;
ssize_t y = ...;
printf("%zu\n", x);  // prints as unsigned decimal
printf("%zx\n", x);  // prints as hex
printf("%zd\n", y);  // prints as signed decimal
```
These will adapt automatically to the platform dependent size of size_t.

**Review by dkgroot — CHANGES_REQUESTED** [review]
@ibuclaw Thanks a lot for these changes !

File: src/ctfeexpr.c (hunk 1 above)
Comment (dkgroot) [reply src/ctfeexpr.c]: @ibuclaw: I did the same kind of fixes you did (using %zd in my case), just moments before you posted this PR. If you like i can attach the diff here.

**Review by wilzbach — APPROVED** [review]
Whenever you are ready.
Do you want to add the appropriate flags to `posix.mak`?
(For `dmd-cxx` being bug fixes only this is quite an effort though)

File: src/ctfeexpr.c (hunk 1 above)
Comment (ibuclaw) [reply src/ctfeexpr.c]: @dkgroot - Lets just stick to `%l` or `%ll`.  The explicit casts are just a formality anyway.

File: src/expression.c [hunk 2]
```diff
@@ -1928,7 +1888,7 @@ bool functionParameters(Loc loc, Scope *sc, TypeFunction *tf,
 
                 /* Declare temporary 'auto __pfx = arg' (needsDtor) or 'auto __pfy = arg' (!needsDtor)
                  */
-                VarDeclaration *tmp = tmp = copyToTemp(0,
+                VarDeclaration *tmp = copyToTemp(0,
```
Comment (ibuclaw) [inline src/expression.c]: The typo's that warnings found. :-)

**Comment by ibuclaw** [comment]
> (For dmd-cxx being bug fixes only this is quite an effort though)

Bug fixes slash gdc mirror. ;-)

File: src/ctfeexpr.c (hunk 1 above)
Comment (dkgroot) [reply src/ctfeexpr.c]: Ok, it was just a suggestion.
Getting those compiler warnings silenced is the main objective

**Comment by ibuclaw** [comment]
@wilzbach - Added warnings as per 18126.  There's a slight tightening of warnings for the frontend sources, and a loosening of the glue sources, that is loosened further with the backend.

Also added `ENABLE_WERROR` and `ENABLE_WFORMAT`.

Tested locally using `ENABLE_WERROR=1` and `ENABLE_WARNINGS=1 ENABLE_WERROR=1` with both g++ and clang++ as the set compiler.

**Comment by ibuclaw** [comment]
@wilzbach - Shall we merge and move on?

**Review by PetarKirov — APPROVED** [review]

**Pull request merged**